### PR TITLE
components: Convert Select to TypeScript

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Converts the SelectControl to TypeScript. Depends on the InputControl conversion.

I did this as part of https://github.com/WordPress/gutenberg/pull/33714 so just splitting it into a separate PR.

## How has this been tested?
Types should pass. There are no runtime changes 🙂 

## Types of changes
Code quality

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/components/src/select-control/index.tsx [hunk 1]
```diff
@@ -16,15 +18,39 @@ import { Icon, chevronDown } from '@wordpress/icons';
...
+	options?: {
+		label: string;
+		value: string;
+		id?: string;
+		disabled?: boolean;
+	}[];
+	size?: Size;
+	value?: string | string[];
+	labelPosition?: LabelPosition;
+}
```
Comment (ciampo) [inline packages/components/src/select-control/index.tsx:52]: Are all of these props supposed to be optional?

File: packages/components/src/select-control/index.tsx [hunk 2]
```diff
@@ -89,15 +115,17 @@ function SelectControl(
 				label={ label }
 				size={ size }
 				suffix={
-					<DownArrowWrapper>
-						<Icon icon={ chevronDown } size={ 18 } />
-					</DownArrowWrapper>
+					props.suffix || (
```
Comment (ciampo) [inline packages/components/src/select-control/index.tsx]: Is passing `props.suffix` here  (and omitting it from the props passed to <Select />) a bug fix? If yes, should it be included in the changelog?

File: packages/components/src/select-control/index.tsx [hunk 3]
```diff
@@ -89,15 +115,17 @@ function SelectControl(
...
+						<DownArrowWrapper>
+							<Icon icon={ chevronDown } size={ 18 } />
+						</DownArrowWrapper>
+					)
 				}
+				prefix={ props.prefix }
```
Comment (ciampo) [inline packages/components/src/select-control/index.tsx]: Is passing `prefix` here (and omitting it from the props passed to `<Select />`) a bug fix? If yes, should it be included in the changelog?

File: packages/components/src/select-control/styles/select-control-styles.ts [hunk 4]
```diff
@@ -75,7 +81,7 @@ export const Select = styled.select`
 		${ fontSizeStyles };
 		${ sizeStyles };
 
-		${ rtl( { paddingLeft: 8, paddingRight: 24 } )() }
+		${ rtl( { paddingLeft: 8, paddingRight: 24 } ) }
```
Comment (ciampo) [inline packages/components/src/select-control/styles/select-control-styles.ts:84]: Were rtl styles previously not correctly applied correctly? If yes, should this bug fix be included in the changelog?

File: packages/components/src/select-control/index.tsx [hunk 5]
```diff
@@ -89,15 +115,17 @@ function SelectControl(
...
 				labelPosition={ labelPosition }
-				{ ...props }
 			>
 				<Select
-					{ ...props }
+					{ ...omit( props, 'prefix', 'suffix' ) }
```
Comment (ciampo) [inline packages/components/src/select-control/index.tsx]: Instead of omitting `prefix` and `suffix`, could we just destructure them at the start of the render function?

File: packages/components/src/select-control/index.tsx (hunk 1 above)
Comment (sarayourfriend) [reply packages/components/src/select-control/index.tsx:52]: They all have default values or default handling so yes, I think so. Did you notice any in particular that you think should not be optional?

File: packages/components/src/select-control/index.tsx (hunk 2 above)
Comment (sarayourfriend) [reply packages/components/src/select-control/index.tsx]: It's not really a bug fix... there was no bug being caused earlier, but it was causing it impossible to truly document the full types of the component because we were spreading `props` into two places. By being explicit we can prove to TypeScript that the only props we spread into `InputBase` are truly the ones that it should have.

File: packages/components/src/select-control/index.tsx (hunk 3 above)
Comment (sarayourfriend) [reply packages/components/src/select-control/index.tsx]: Ditto here, I don't think it qualifies as a bug fix as it's just to support better typing. There was no bug before.

File: packages/components/src/select-control/styles/select-control-styles.ts (hunk 4 above)
Comment (sarayourfriend) [reply packages/components/src/select-control/styles/select-control-styles.ts:84]: I think they just weren't responsive to other prop changes (but not to `isRtl` changes). I have no idea if this was causing a bug, I'm actually fairly certain that the `rtl` function doesn't really work the way it is intended to work (that is responsively to changes in the `isRtl()` return value... it will only actually update if the component re-renders but nothing in the `rtl` mixin will actually do that.

We might need to do something like create a `withRtl` wrapper to wrap styled components that uses a hook to watch the `RTL` value and re-render appropriately. Right now it'll only update if non-related things cause the component to re-render.

In short, no bug fix here really as the rtl utility doesn't even work correctly to begin with (I think).

Comment (sarayourfriend) [reply packages/components/src/select-control/styles/select-control-styles.ts:84]: I'm going to open a PR to explore this issue further, BTW.

File: packages/components/src/select-control/index.tsx (hunk 2 above)
Comment (ciampo) [reply packages/components/src/select-control/index.tsx]: Got it, thank you!

File: packages/components/src/select-control/styles/select-control-styles.ts (hunk 4 above)
Comment (ciampo) [reply packages/components/src/select-control/styles/select-control-styles.ts:84]: Thanks for the explanation and for looking further into this. Not a blocker for this PR

File: packages/components/src/select-control/index.tsx (hunk 1 above)
Comment (ciampo) [reply packages/components/src/select-control/index.tsx:52]: > Did you notice any in particular that you think should not be optional?

No, I didn't. It just felt unusual at a first glance, but as you said, all props seem to either be handled or passed to subcomponents for which they are also optional.

**Review by ciampo — APPROVED** [review]
LGTM 🚀

File: packages/components/src/select-control/index.tsx [hunk 6]
```diff
@@ -1,8 +1,10 @@
 /**
  * External dependencies
  */
-import { isEmpty, noop } from 'lodash';
+import { isEmpty, noop, omit } from 'lodash';
```
Comment (ciampo) [inline packages/components/src/select-control/index.tsx:4]: Just realised this — can the `omit` import be removed now? I don't see any further usages.
We may want to apply this change directly in #33696
